### PR TITLE
implement context exchange points

### PR DESCRIPTION
**Pull request by Robbbo-T**
## Title
_Replace this with the PR title_

## Summary
- Add `.github/FUNDING.yml` with `github: ['Robbbo-T']` and custom fallback
- Add `FINANCE/SPONSORSHIP.md` (IDEALE-EU, MAL-EEM, anti-speculative)
- Wire funding docs into `utcs/manifest.yaml`
- Add Sponsors badge to README
- CI sanity check for FUNDING.yml

## Canon checks
- [x] Bridge QS→FWD→UE→FE→CB→QB unchanged
- [x] UTCS phrasing intact
- [x] QOx = Quantum Optimizations
- [x] PAx orientation markers: OB/OFF only
- [x] AMPEL360 remains commons (no app code)

## Evidence
- [ ] CI `Funding config check` passes
- [ ] UTCS bundle includes `FINANCE/SPONSORSHIP.md`

## Summary by Sourcery

Integrate the Teknia Token (TT) system v3.14 into the repository with a new Python CLI, hybrid π-tier tokenomics, comprehensive documentation, test suites, and CI workflows.

New Features:
- Add `tools/tek_tokens.py` CLI for ledger initialization, transfers, rewards, consumption, fee quoting, verification, and badge generation with integer ‘deg’ precision.
- Implement π-tier transfer fees (0.314%, 0.99%, 3.14%) alongside a base 0.5% fee for reward/consume operations.
- Enforce 360° divisibility (1 TT=360 deg) and a minimum transfer quantum of 2592 deg on transfers only.
- Introduce policy immutability via SHA-256 hash stored in `finance/ledger.json` and an append-only transaction hash chain (`txlog.jsonl`/`txhead.json`).
- Provide two tokenomics configurations: a hybrid fee schedule and a no-fee variant.

CI:
- Introduce GitHub Actions workflow `.github/workflows/tek-tokens-verify.yml` to validate tokenomics config, CLI functionality, ledger integrity, and badge generation.

Documentation:
- Add comprehensive user documentation in `docs/TOKENS.md`, `finance/README.md`, `tools/README.md`, `docs/README.md`, and update `README-PAGES.md` to surface the token system.
- Include Copilot guidance in `.github/copilot-instructions.md` for maintaining tokenomics consistency.

Tests:
- Add v3.1 and v3.14 test suites (`tools/test_tek_tokens.py` and `tools/test_tek_tokens_v314.py`) covering initialization, fee calculations, quantum enforcement, policy checks, CLI outputs, and badge generation.

Chores:
- Update `.gitignore` to exclude generated ledger, log, head, and badge artifacts.

### CODE REVIEW

File: tools/test_tek_tokens_v314.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    
+    # --- helpers for robust stdout checks ---
+    @staticmethod
+    def _has_number_with_unit(text: str, value: int, unit: str = "deg") -> bool:
+        """
+        True if 'text' contains the integer 'value' followed by 'unit', with or without
+        thousands separators (e.g., '2,566 deg' or '2566 deg').
+        """
+        pattern = rf"(?:\b{value:,}\s*{unit}\b|\b{value}\s*{unit}\b)"
+        return re.search(pattern, text) is not None
```
[the remark this answers is not in the document]
Comment (Robbbo-T) [reply tools/test_tek_tokens_v314.py:99]: @copilot diff --git a/tools/test_tek_tokens_v314.py b/tools/test_tek_tokens_v314.py
@@
     @staticmethod
     def _has_number_with_unit(text: str, value: int, unit: str = "deg") -> bool:
         """
-        True if 'text' contains the integer 'value' followed by 'unit', with or without
-        thousands separators (e.g., '2,566 deg' or '2566 deg').
+        True if 'text' contains the integer 'value' followed by 'unit', with or without
+        thousands separators (e.g., '2,566 deg' or '2566 deg'). Accepts comma or common
+        thin/nb space separators defensively.
         """
-        pattern = rf"(?:\b{value:,}\s*{unit}\b|\b{value}\s*{unit}\b)"
+        # Accept: comma, narrow no-break space, non-breaking space
+        sep = r"(?:,|\u202F|\u00A0)"
+        # Build a variant of value with flexible separators (e.g., 81,388 -> 81(?:,|\u202F|\u00A0)388)
+        value_with_sep = f"{value:,}".replace(",", "{SEP}")
+        value_with_sep = value_with_sep.replace("{SEP}", sep)
+        pattern = rf"(?:\b{value_with_sep}\s*{unit}\b|\b{value}\s*{unit}\b)"
         return re.search(pattern, text) is not None

File: tools/tek_tokens.py [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+        sorted_tiers = sorted(schedule, key=lambda t: t["min_deg"], reverse=True)
+        
+        for tier in sorted_tiers:
+            if amount_deg >= tier["min_deg"]:
+                # Parse BPS (might be string like "31.4" or int like 99)
+                bps_val = tier["bps"]
+                # Always use Fraction for consistent calculation
+                bps_fraction = Fraction(bps_val)
+                # Calculate fee: amount * (bps/10000), floored, using exact Fraction arithmetic
+                fee = int((amount_deg * bps_fraction) // 10000)
```
[the remark this answers is not in the document]
Comment (Robbbo-T) [reply tools/tek_tokens.py:135]: @copilot -                bps_fraction = Fraction(bps_val)
-                # Calculate fee: amount * (bps/10000), floored, using exact Fraction arithmetic
-                fee = int((amount_deg * bps_fraction) // 10000)
+                # Parse BPS deterministically (avoid float artifacts like Fraction(31.4))
+                bps_fraction = Fraction(str(bps_val))
+                # fee = floor(amount_deg * (bps / 10000)) using pure integer arithmetic
+                fee = (amount_deg * bps_fraction.numerator) // (10000 * bps_fraction.denominator)

File: tools/test_tek_tokens_v314.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+        self.assert_true(result.returncode == 0, "Badge command succeeded")
+        
+        badge_file = REPO_ROOT / "badges" / "tt-verified.svg"
+        self.assert_true(badge_file.exists(), "Badge SVG file created")
+        
+        with badge_file.open() as f:
+            content = f.read()
+        self.assert_true("<svg" in content, "Badge is valid SVG")
+    
+    def test_eur_valuation(self):
```
[the remark this answers is not in the document]
Comment (Robbbo-T) [reply tools/test_tek_tokens_v314.py:241]: @copilot diff --git a/tools/test_tek_tokens_v314.py b/tools/test_tek_tokens_v314.py
index b1a2e3c..ad3d8e1 100755
--- a/tools/test_tek_tokens_v314.py
+++ b/tools/test_tek_tokens_v314.py
@@ -17,6 +17,7 @@ from pathlib import Path
 import tempfile
 import shutil
 import subprocess
+import math
 
 # Add tools directory to path
 sys.path.insert(0, str(Path(__file__).parent))
@@ -100,6 +101,37 @@ class TestTokenLedgerV314:
         pattern = rf"(?:\b{value:,}\s*{unit}\b|\b{value}\s*{unit}\b)"
         return re.search(pattern, text) is not None
 
+    # --- helpers for robust EUR checks ---
+    @staticmethod
+    def _eur_patterns_for(value: float, decimals: int = 2):
+        """
+        Build regex-safe alternatives for displaying a EUR amount with:
+        - optional thousands separators
+        - either '.' or ',' as decimal separator
+        """
+        rounded = f"{value:.{decimals}f}"             # e.g., '72.66'
+        dot = rounded                                 # '72.66'
+        comma = rounded.replace(".", ",")             # '72,66'
+        # With thousands (unlikely for these small numbers, but supported):
+        dot_thousands = f"{float(rounded):,.{decimals}f}"  # '1,234.56'
+        comma_thousands = dot_thousands.replace(",", "X").replace(".", ",").replace("X", ".")  # '1.234,56'
+        return {dot, comma, dot_thousands, comma_thousands}
+
+    @classmethod
+    def _has_eur_amount(cls, text: str, value: float, decimals: int = 2) -> bool:
+        """
+        True if 'text' contains a '€' value that equals 'value' to 'decimals'
+        (supports optional thousands separators and '.' or ',').
+        """
+        alts = cls._eur_patterns_for(value, decimals=decimals)
+        # Build a single regex that matches any of the formatted variants
+        alt_union = "|".join(re.escape(a) for a in alts)
+        pattern = rf"€\s*(?:{alt_union})\b"
+        return re.search(pattern, text) is not None
+
+    @staticmethod
+    def _deg_to_tt(deg: int) -> float:
+        return deg / 360.0
+
     def run_cli(self, *args):
         """Run CLI command and return result."""
         cmd = [sys.executable, str(REPO_ROOT / "tools" / "tek_tokens.py")] + list(args)
@@ -261,13 +293,41 @@ class TestTokenLedgerV314:
     def test_eur_valuation(self):
         """Test EUR valuation display."""
         print("\nTest 8: EUR Valuation")
 
         ledger_file = REPO_ROOT / "finance" / "ledger.json"
         if not ledger_file.exists():
             self.run_cli("init")
 
-        result = self.run_cli("--eur-per-tt", "0.10", "quote", "--op", "transfer", "--tt", "720")
+        # Quote 720 TT transfer at €0.10 / TT
+        result = self.run_cli("--eur-per-tt", "0.10", "quote", "--op", "transfer", "--tt", "720")
         self.assert_true(result.returncode == 0, "EUR valuation succeeded")
-        self.assert_true("€" in result.stdout, "EUR symbol displayed")
+        self.assert_true("€" in result.stdout, "EUR symbol displayed")
+
+        # Expected amounts:
+        # amount_deg = 720 * 360 = 259,200
+        # tier fee (0.99%) in deg = floor(259,200 * 99 / 10,000) = 2,566
+        amount_deg = 720 * 360
+        fee_deg = (amount_deg * 99) // 10000
+        sender_deg = amount_deg + fee_deg
+
+        eur_per_tt = 0.10
+        amount_eur = TestTokenLedgerV314._deg_to_tt(amount_deg) * eur_per_tt      # 720 TT * 0.10 = 72.00
+        fee_eur = TestTokenLedgerV314._deg_to_tt(fee_deg) * eur_per_tt            # (2566/360) * 0.10 ≈ 0.71
+        sender_eur = TestTokenLedgerV314._deg_to_tt(sender_deg) * eur_per_tt      # ≈ 72.66
+
+        # Check that recipient (the nominal transfer amount) shows as €72.00
+        self.assert_true(
+            self._has_eur_amount(result.stdout, amount_eur, decimals=2),
+            "Recipient EUR matches expected (€72.00 for 720 TT @ €0.10/TT)"
+        )
+
+        # Check that sender pays (amount + fee) shows as ~€72.66
+        self.assert_true(
+            self._has_eur_amount(result.stdout, sender_eur, decimals=2),
+            "Sender EUR matches expected (amount + 0.99% tier fee)"
+        )
+
+        # If a Fee line is present, also validate the fee EUR (~€0.71)
+        if "Fee" in result.stdout or "fee" in result.stdout:
+            self.assert_true(self._has_eur_amount(result.stdout, fee_eur, decimals=2),
+                             "Fee EUR matches expected (~€0.71)")
 
     def test_balance_verification(self):
         """Test that total balance equals genesis supply after transfers."""
         print("\nTest 9: Balance Conservation")

File: tools/test_tek_tokens_v314.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        True if 'text' contains the integer 'value' followed by 'unit', with or without
+        thousands separators (e.g., '2,566 deg' or '2566 deg').
+        """
+        pattern = rf"(?:\b{value:,}\s*{unit}\b|\b{value}\s*{unit}\b)"
+        return re.search(pattern, text) is not None
+    
+    def run_cli(self, *args):
+        """Run CLI command and return result."""
+        cmd = [sys.executable, str(REPO_ROOT / "tools" / "tek_tokens.py")] + list(args)
+        result = subprocess.run(cmd, capture_output=True, text=True)
```
[the remark this answers is not in the document]
Comment (Robbbo-T) [reply tools/test_tek_tokens_v314.py:104]: @copilot diff --git a/tools/test_tek_tokens_v314.py b/tools/test_tek_tokens_v314.py
index ad3d8e1..c5b9a8d 100755
--- a/tools/test_tek_tokens_v314.py
+++ b/tools/test_tek_tokens_v314.py
@@ -112,10 +112,17 @@ class TestTokenLedgerV314:
         return re.search(pattern, text) is not None
 
     def run_cli(self, *args):
         """Run CLI command and return result."""
-        cmd = [sys.executable, str(REPO_ROOT / "tools" / "tek_tokens.py")] + list(args)
-        result = subprocess.run(cmd, capture_output=True, text=True)
+        # NOTE: We pass argv as a list with shell=False (explicit) to avoid any shell parsing.
+        # Args are provided by the test suite (not external input).
+        cmd = [sys.executable, str(REPO_ROOT / "tools" / "tek_tokens.py")] + [str(a) for a in args]
+        result = subprocess.run(
+            cmd,
+            shell=False,               # explicit: no shell -> no shell injection vector
+            capture_output=True,
+            text=True,
+            cwd=str(REPO_ROOT),        # stable working directory
+        )
         return result

File: tools/test_tek_tokens.py [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+import shutil
+
+# Add tools directory to path
+sys.path.insert(0, str(Path(__file__).parent))
+
+# Import the token ledger
+from tek_tokens import TokenLedger, TT_TO_DEG, GENESIS_SUPPLY_DEG, GENESIS_SUPPLY_TT
+
+
+class TestTokenLedger:
```
[the remark this answers is not in the document]
Comment (Robbbo-T) [reply tools/test_tek_tokens.py:21]: @copilot diff --git a/tools/test_tek_tokens_v314.py b/tools/test_tek_tokens_v314.py
index 8f6f3a1..a1b9e5c 100755
--- a/tools/test_tek_tokens_v314.py
+++ b/tools/test_tek_tokens_v314.py
@@ -26,6 +26,7 @@ import subprocess
 
 # Add tools directory to path
 sys.path.insert(0, str(Path(__file__).parent))
+ERR = lambda s: s.lower().find("error") >= 0  # tiny helper
 
 REPO_ROOT = Path(__file__).parent.parent
 
@@ -88,6 +89,18 @@ class TestTokenLedgerV314:
             print(f"  ✗ {message}")
             self.tests_failed += 1
 
+    # --- tiny helpers for error/string checks ---
+    @staticmethod
+    def _stream_contains(result, *needles: str) -> bool:
+        hay = (result.stdout or "") + "\n" + (result.stderr or "")
+        hay_low = hay.lower()
+        return all(n.lower() in hay_low for n in needles)
+
+    @staticmethod
+    def _has_json_error(result) -> bool:
+        # robust to different json error messages across python versions
+        return any(k in ((result.stdout or "") + result.stderr).lower() for k in ["json", "decode", "expecting", "parse", "invalid"])
+
     # --- helpers for robust stdout checks ---
     @staticmethod
     def _has_number_with_unit(text: str, value: int, unit: str = "deg") -> bool:
@@ -264,6 +277,116 @@ class TestTokenLedgerV314:
         genesis_deg = 720_000_000_000
         self.assert_true(total == genesis_deg, f"Total balance equals genesis ({total:,} == {genesis_deg:,})")
 
+    # ------------------------------
+    # Error handling & robustness
+    # ------------------------------
+    def test_missing_config_file_cli(self):
+        """CLI should fail clearly when an explicit --config path is missing."""
+        print("\nTest 10: Missing Config File (CLI)")
+        with tempfile.TemporaryDirectory() as tmpdir:
+            missing_cfg = Path(tmpdir) / "nope_config.json"
+            # Explicitly pass nonexistent config
+            result = self.run_cli("--config", str(missing_cfg), "init")
+            self.assert_true(result.returncode != 0, "Init fails with missing config file")
+            self.assert_true(ERR(result.stderr) or ERR(result.stdout) or
+                             self._stream_contains(result, "not found") or
+                             self._stream_contains(result, "no such file"),
+                             "Error message mentions missing file")
+
+    def test_corrupted_config_file_cli(self):
+        """CLI should fail clearly when the config file is corrupted."""
+        print("\nTest 11: Corrupted Config File (CLI)")
+        with tempfile.TemporaryDirectory() as tmpdir:
+            bad_cfg = Path(tmpdir) / "corrupted_config.json"
+            bad_cfg.write_text("{not: valid json", encoding="utf-8")
+            result = self.run_cli("--config", str(bad_cfg), "init")
+            self.assert_true(result.returncode != 0, "Init fails with corrupted config file")
+            self.assert_true(self._has_json_error(result), "Error mentions JSON/parse/decode problem")
+
+    def test_missing_ledger_file_cli(self):
+        """CLI should fail clearly when the ledger file is missing for verify/balance."""
+        print("\nTest 12: Missing Ledger File (CLI)")
+        # Ensure clean state
+        ledger_file = REPO_ROOT / "finance" / "ledger.json"
+        if ledger_file.exists():
+            ledger_file.unlink()
+        # Running verify without a ledger should fail
+        result = self.run_cli("verify")
+        self.assert_true(result.returncode != 0, "Verify fails when ledger file is missing")
+        self.assert_true(ERR(result.stderr) or ERR(result.stdout) or
+                         self._stream_contains(result, "ledger") or
+                         self._stream_contains(result, "not found"),
+                         "Error mentions missing ledger")
+
+    def test_corrupted_ledger_file_cli(self):
+        """CLI should fail clearly when the ledger file is corrupted."""
+        print("\nTest 13: Corrupted Ledger File (CLI)")
+        # Create a valid ledger first
+        self.run_cli("init")
+        ledger_file = REPO_ROOT / "finance" / "ledger.json"
+        # Corrupt it
+        ledger_file.write_text("{not: valid json", encoding="utf-8")
+        # verify/balance should now fail
+        result = self.run_cli("verify")
+        self.assert_true(result.returncode != 0, "Verify fails with corrupted ledger file")
+        self.assert_true(self._has_json_error(result), "Error mentions JSON/parse/decode problem")
+
     def run_all_tests(self):
         """Run all tests."""
         print("="*60)
         print("Teknia Token CLI Test Suite v3.14")
         print("Testing π-tier fees, policy hash, and enhanced features")
         print("="*60)
@@ -279,6 +402,10 @@ class TestTokenLedgerV314:
             self.test_badge_command()
             self.test_eur_valuation()
             self.test_balance_verification()
+            # Error handling / robustness
+            self.test_missing_config_file_cli()
+            self.test_corrupted_config_file_cli()
+            self.test_missing_ledger_file_cli()
+            self.test_corrupted_ledger_file_cli()
 
         finally:
             self.teardown()

File: tools/test_tek_tokens_v314.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+        
+        # Check founder allocation (5%)
+        genesis_deg = 720_000_000_000
+        founder_deg = (genesis_deg * 500) // 10000
+        self.assert_true(
+            ledger["accounts"]["FOUNDER"] == founder_deg,
+            f"Founder has 5% allocation ({founder_deg:,} deg)"
+        )
+    
+    def test_pi_tier_fees(self):
```
[the remark this answers is not in the document]
Comment (Robbbo-T) [reply tools/test_tek_tokens_v314.py:136]: @copilot diff --git a/tools/test_tek_tokens_v314.py b/tools/test_tek_tokens_v314.py
index 4a3b2d1..9c8a0f4 100755
--- a/tools/test_tek_tokens_v314.py
+++ b/tools/test_tek_tokens_v314.py
@@ -255,6 +255,71 @@ class TestTokenLedgerV314:
         self.assert_true(total == genesis_deg, f"Total balance equals genesis ({total:,} == {genesis_deg:,})")
 
     def run_all_tests(self):
+        """Run all tests."""
+        print("="*60)
+        print("Teknia Token CLI Test Suite v3.14")
+        print("Testing π-tier fees, policy hash, and enhanced features")
+        print("="*60)
+
+        try:
+            self.setup()
+
+            self.test_initialization()
+            self.test_pi_tier_fees()
+            self.test_reward_base_fee()
+            self.test_min_transfer_scope()
+            self.test_quote_command()
+            self.test_verify_command()
+            self.test_badge_command()
+            self.test_eur_valuation()
+            self.test_balance_verification()
+
+        finally:
+            self.teardown()
+
+        return self.tests_failed == 0
+
+    def test_pi_tier_boundaries(self):
+        """Boundary tests for π-tier fee calculation at and around thresholds.
+        Uses multiples of 7.2 TT to satisfy min_transfer_deg (2592 deg) constraint.
+        Tiers:
+          - < 72 TT: base 0.5% (50 bps)
+          - ≥ 72 TT: 0.314% (31.4 bps)
+          - ≥ 720 TT: 0.99% (99 bps)
+          - ≥ 7200 TT: 3.14% (314 bps)
+        Fee is floor(amount_deg * rate).
+        """
+        print("\nTest X: π-Tier Boundary Conditions")
+
+        ledger_file = REPO_ROOT / "finance" / "ledger.json"
+        if not ledger_file.exists():
+            self.run_cli("init")
+
+        # Helper to compute expected fee in deg via integer math
+        def fee_deg(amount_tt: float, num: int, den: int) -> int:
+            amount_deg = int(round(amount_tt * 360))
+            return (amount_deg * num) // den
+
+        cases = [
+            # Below first threshold (uses base 0.5%): 64.8 TT = 9×7.2
+            ("64.8", 50, 10000, 116),
+            # At first threshold (0.314%): 72.0 TT = 10×7.2
+            ("72.0", 314, 100000, 81),
+            # Just above first threshold (0.314%): 79.2 TT = 11×7.2
+            ("79.2", 314, 100000, 89),
+            # Just below second threshold (0.314%): 712.8 TT = 99×7.2
+            ("712.8", 314, 100000, 805),
+            # At second threshold (0.99%): 720.0 TT = 100×7.2
+            ("720.0", 99, 10000, 2566),
+            # Just above second threshold (0.99%): 727.2 TT = 101×7.2
+            ("727.2", 99, 10000, 2591),
+            # Just below third threshold (0.99%): 7192.8 TT = 999×7.2
+            ("7192.8", 99, 10000, 25635),
+            # At third threshold (3.14%): 7200.0 TT = 1000×7.2
+            ("7200.0", 314, 10000, 81388),
+            # Just above third threshold (3.14%): 7207.2 TT = 1001×7.2
+            ("7207.2", 314, 10000, 81470),
+        ]
+
+        for tt_str, num, den, expected in cases:
+            result = self.run_cli("transfer", "--from", "TREASURY", "--to", "boundary", "--tt", tt_str)
+            self.assert_true(result.returncode == 0, f"{tt_str} TT transfer succeeded")
+            # Double-check computed fee and assert it appears in stdout (robust to comma formatting)
+            calc = fee_deg(float(tt_str), num, den)
+            self.assert_true(calc == expected, f"Computed fee matches expected for {tt_str} TT ({calc} == {expected})")
+            self.assert_true(self._has_number_with_unit(result.stdout, expected, "deg"),
+                             f"Output shows expected fee {expected:,} deg for {tt_str} TT")
+
-        """Run all tests."""
-        print("="*60)
-        print("Teknia Token CLI Test Suite v3.14")
-        print("Testing π-tier fees, policy hash, and enhanced features")
-        print("="*60)
-
-        try:
-            self.setup()
-
-            self.test_initialization()
-            self.test_pi_tier_fees()
-            self.test_reward_base_fee()
-            self.test_min_transfer_scope()
-            self.test_quote_command()
-            self.test_verify_command()
-            self.test_badge_command()
-            self.test_eur_valuation()
-            self.test_balance_verification()
-
-        finally:
-            self.teardown()
-
-        return self.tests_failed == 0
+        print("  ✓ π-tier boundary conditions validated")
 
 def main():
     """Run test suite."""

File: tools/test_tek_tokens_v314.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        
+        # Initialize if not already done
+        ledger_file = REPO_ROOT / "finance" / "ledger.json"
+        if not ledger_file.exists():
+            self.run_cli("init")
+        
+        # Test 72 TT transfer (should use 0.314% tier)
+        result = self.run_cli("transfer", "--from", "TREASURY", "--to", "alice", "--tt", "72")
+        self.assert_true(result.returncode == 0, "72 TT transfer succeeded")
+        self.assert_true("81 deg" in result.stdout, "Fee is 81 deg (0.314% of 25920)")
```
[the remark this answers is not in the document]
Comment (Robbbo-T) [reply tools/test_tek_tokens_v314.py:148]: diff --git a/tools/test_tek_tokens_v314.py b/tools/test_tek_tokens_v314.py
--- a/tools/test_tek_tokens_v314.py
+++ b/tools/test_tek_tokens_v314.py
@@ -168,7 +168,8 @@ class TestTokenLedgerV314:
         # Test 72 TT transfer (should use 0.314% tier)
         result = self.run_cli("transfer", "--from", "TREASURY", "--to", "alice", "--tt", "72")
         self.assert_true(result.returncode == 0, "72 TT transfer succeeded")
-        self.assert_true("81 deg" in result.stdout, "Fee is 81 deg (0.314% of 25920)")
+        self.assert_true(self._has_number_with_unit(result.stdout, 81, "deg"),
+                         "Fee is 81 deg (0.314% of 25920)")
 
         # Test 720 TT transfer (should use 0.99% tier)
         result = self.run_cli("transfer", "--from", "TREASURY", "--to", "bob", "--tt", "720")

File: tools/test_tek_tokens_v314.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+        self.assert_true(result.returncode == 0, "Quote command succeeded")
+        self.assert_true(self._has_number_with_unit(result.stdout, 2566, "deg"),
+                        "Quote shows correct fee (0.99% tier)")
+        self.assert_true("Sender pays:" in result.stdout, "Quote shows sender cost")
+        self.assert_true("Recipient gets:" in result.stdout, "Quote shows recipient amount")
+        
+        # Quote for reward should show 0.5% fee
+        result = self.run_cli("quote", "--op", "reward", "--tt", "72")
+        self.assert_true(result.returncode == 0, "Quote reward succeeded")
+        self.assert_true("129 deg" in result.stdout, "Quote shows 0.5% fee for reward")
```
[the remark this answers is not in the document]
Comment (Robbbo-T) [reply tools/test_tek_tokens_v314.py:206]: diff --git a/tools/test_tek_tokens_v314.py b/tools/test_tek_tokens_v314.py
--- a/tools/test_tek_tokens_v314.py
+++ b/tools/test_tek_tokens_v314.py
@@ -196,7 +196,8 @@ class TestTokenLedgerV314:
         # Test 72 TT reward (should use 0.5% = 129 deg fee, not 0.314% = 81 deg)
         result = self.run_cli("reward", "--to", "dave", "--tt", "72")
         self.assert_true(result.returncode == 0, "72 TT reward succeeded")
-        self.assert_true("129 deg" in result.stdout, "Fee is 129 deg (0.5% of 25920)")
+        self.assert_true(self._has_number_with_unit(result.stdout, 129, "deg"),
+                         "Fee is 129 deg (0.5% of 25920)")
 
     def test_min_transfer_scope(self):
         """Test that min_transfer_deg only applies to transfers, not reward/consume."""

**Review by Robbbo-T — COMMENTED** [review]
@copilot This is excellent work. v3.14 is coherently wired through config, docs, CLI, tests, and CI. I did a quick repo-wide review and here are the few surgical tweaks I’d make to tighten things up and keep everything deterministic and secure.

# Top nits & fixes

1. **Badge path mismatch in CI**

* Your workflow checks for `finance/token_badge.json`, but the CLI (per docs/tests) emits an SVG at `badges/tt-verified.svg`. Either have the CLI also emit a JSON summary, or—simpler—update the workflow to check the SVG.

```diff
diff --git a/.github/workflows/tek-tokens-verify.yml b/.github/workflows/tek-tokens-verify.yml
@@ -60,13 +60,13 @@
       - name: Verify badge generation
         run: |
           echo "=== Testing Badge Generation ==="
           if [ -f finance/ledger.json ]; then
             python3 tools/tek_tokens.py verify
-            if [ -f finance/token_badge.json ]; then
-              echo "✓ Badge generated successfully"
-              cat finance/token_badge.json
+            if [ -f badges/tt-verified.svg ]; then
+              echo "✓ Badge SVG generated successfully"
+              ls -l badges/tt-verified.svg
             else
               echo "✗ Badge generation failed"
               exit 1
             fi
           else
```

2. **Deterministic JSON writes**

* For any JSON you write (ledger, txlog entries that are JSON objects, etc.), use canonical separators and sorted keys to keep diffs stable and verification reproducible.

```diff
diff --git a/tools/tek_tokens.py b/tools/tek_tokens.py
@@ -xxx,7 +xxx,12 @@ def save_ledger(ledger: dict, path: Path = LEDGER):
-    with path.open("w", encoding="utf-8", newline="\n") as f:
-        json.dump(ledger, f, ensure_ascii=False)
+    with path.open("w", encoding="utf-8", newline="\n") as f:
+        json.dump(
+            ledger,
+            f,
+            ensure_ascii=False,
+            sort_keys=True,
+            separators=(",", ":"),
+        )
```

(Apply the same pattern anywhere else you `json.dump`.)

3. **SVG text escaping (security hygiene)**

* You already flagged this in review comments—good call. Here’s a concrete, minimal helper and usage:

```diff
diff --git a/tools/tek_tokens.py b/tools/tek_tokens.py
@@ -480,6 +480,10 @@
+import html
+
+def _svg_escape(s: object) -> str:
+    return html.escape(str(s), quote=False)
@@ -493,9 +497,9 @@ def badge(...):
-    svg = f'''<svg xmlns="http://www.w3.org/2000/svg" width="200" height="40">
+    svg = f'''<svg xmlns="http://www.w3.org/2000/svg" width="200" height="40">
   <rect width="200" height="40" fill="#0366d6"/>
   <text x="100" y="25" font-family="Arial" font-size="16" fill="white" text-anchor="middle">
-    TT: {treasury_tt:,.0f}
+    TT: {_svg_escape(f"{treasury_tt:,.0f}")}
   </text>
 </svg>'''
```

4. **Fee calculation: stick to integer floor**

* Your Fraction-based computation is good; make sure the denominator is included in the divisor to avoid accidental float work anywhere:

```diff
diff --git a/tools/tek_tokens.py b/tools/tek_tokens.py
@@ -420,7 +420,10 @@ def _compute_fee(amount_deg: int, bps_val: str|int) -> int:
-    bps_fraction = Fraction(bps_val)
-    fee = int((amount_deg * bps_fraction) // 10000)
+    bps_fraction = Fraction(bps_val)  # accepts "31.4" etc.
+    # fee = floor(amount_deg * (bps/10000))
+    fee = (amount_deg * bps_fraction.numerator) // (10000 * bps_fraction.denominator)
     return fee
```

5. **Hash-chain completeness**

* Include **all** relevant fields (including `fee`, `tier_label`, etc.) in the tx hash payload, and serialize canonically before hashing. That way audits can reproduce the chain exactly.

```diff
diff --git a/tools/tek_tokens.py b/tools/tek_tokens.py
@@ -394,13 +394,26 @@ def log_tx(tx: dict, policy_hash: Optional[str]):
-    with TXLOG.open('a') as f:
-        f.write(json.dumps(tx) + '\n')
+    # Canonicalize tx entry for hashing & logging
+    canon = json.dumps(tx, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
+    with TXLOG.open('a', encoding="utf-8", newline="\n") as f:
+        f.write(canon + "\n")
 
     # Update txhead with hash chain
     prev_hash = ""
     if TXHEAD.exists():
-        with TXHEAD.open() as f:
+        with TXHEAD.open(encoding="utf-8") as f:
             prev_hash = f.read().strip()
-    head = sha256((prev_hash + canon).encode("utf-8")).hexdigest()
+    head = sha256((prev_hash + canon).encode("utf-8")).hexdigest()
     TXHEAD.parent.mkdir(parents=True, exist_ok=True)
-    with TXHEAD.open('w') as f:
-        f.write(head)
+    with TXHEAD.open('w', encoding="utf-8", newline="\n") as f:
+        f.write(head)
```

6. **Tests: resilient numeric assertions**

* You already introduced `_has_number_with_unit`; nice. You can extend it to also accept thin/nbsp separators (you did in a follow-up). Keep using it consistently (you’ve updated most call sites).

7. **Tests: add boundary checks for π-tiers**

* Add cases at exactly 72 TT, 720 TT, 7,200 TT and one less than each to confirm tier selection changes at the correct thresholds (you referenced this—worth landing as a short PR to lock behavior).

8. **CI: actually run the v3.14 test suite**

* The workflow currently verifies JSON and basic CLI help. Add a job step to execute `tools/test_tek_tokens_v314.py` so regressions can’t sneak in.

```diff
diff --git a/.github/workflows/tek-tokens-verify.yml b/.github/workflows/tek-tokens-verify.yml
@@ -18,7 +18,7 @@ jobs:
   verify-tokens:
     runs-on: ubuntu-latest
 
     steps:
       - name: Checkout repository
-        uses: actions/checkout@v4
+        uses: actions/checkout@v4
@@ -26,7 +26,7 @@ jobs:
       - name: Set up Python
-        uses: actions/setup-python@v4
+        uses: actions/setup-python@v5
         with:
-          python-version: '3.x'
+          python-version: '3.11'
@@ -70,6 +70,15 @@ jobs:
           fi
 
+      - name: Run v3.14 test suite
+        run: |
+          echo "=== Running v3.14 tests ==="
+          python3 tools/test_tek_tokens_v314.py
+        env:
+          PYTHONUTF8: "1"
+          LC_ALL: C.UTF-8
+          LANG: C.UTF-8
+
```

# Optional polish (nice-to-haves)

* **`--json` output mode** for commands like `balance`, `quote`, `verify` to make assertions simpler in tests and enable downstream tooling. (You can land this later; not a blocker.)
* **`subprocess.run` warning in tests**: It’s safe as you build the arg list directly and don’t shell=True. You can add a comment to silence automated scanners.

```diff
-        result = subprocess.run(cmd, capture_output=True, text=True)
+        # Safe: no shell=True; args are a list we control.
+        result = subprocess.run(cmd, capture_output=True, text=True)  # noqa: S603
```

* **Docs pass**: I noticed a few lingering v3.1 mentions are already marked as “Previous”. Just ensure there’s no mixed guidance where an old `tx --type` example might remain (looked clean from your snippets).

---

If you want, I can bundle the diffs above into one patch set; otherwise they’re small and easy to cherry-pick. Overall, this is in great shape—tight, auditable, and CI-friendly.
